### PR TITLE
ChainComponents EVM Interface Tests 

### PR DESCRIPTION
**Pull request by silaslenihan**
Description: Changes the ChainReader EVM Interface tests to use the ChainWriter instead of a helper and updates the simulated backend ChainReader tests to implement a TXM + ChainWriter

### Corresponding ticket:
- [BCI-3497](https://smartcontract-it.atlassian.net/browse/BCI-3497)

### Requires Dependencies
- https://github.com/smartcontractkit/chainlink-common/pull/664
- https://github.com/smartcontractkit/chainlink-solana/pull/827

### Resolves Dependencies
- [BCI-3498](https://smartcontract-it.atlassian.net/browse/BCI-3498)
- https://github.com/smartcontractkit/chainlink-evm-testing/pull/151

[BCI-3497]: https://smartcontract-it.atlassian.net/browse/BCI-3497?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[BCI-3498]: https://smartcontract-it.atlassian.net/browse/BCI-3498?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: core/services/relay/evm/chain_reader_test.go
Comment (ilija42) [inline core/services/relay/evm/chain_reader_test.go:238]: I don't really understand what is the point of adding CW into EVM CR tester? CW should be tested separately from CR

File: core/services/relay/evm/chain_reader_test.go
Comment (ilija42) [inline core/services/relay/evm/chain_reader_test.go:3]: I don't really understand what is the point of adding CW into EVM CR tester? CW should be tested separately from CR

**Review by ilija42 — COMMENTED** [review]
I don't really understand what is the point of adding CW into EVM CR tester? CW should be tested separately from CR

File: docs/CONFIG.md
Comment (ilija42) [inline docs/CONFIG.md]: what does this config do?

Comment (silaslenihan) [reply docs/CONFIG.md]: This is used in gas estimation. I'm not sure why it was set so high (100 micro = 100000 gwei) but it was causing the TXM to fail on simulated backend since it would result in astronomically priced TXs that would go over the maximum balance possible on the simulated backend.

File: core/services/relay/evm/chain_components_test.go
Comment (ilija42) [inline core/services/relay/evm/chain_components_test.go]: I would prefer TestChainReader and TestChainWriter to be run as separate test. This need to be either reverted or for `TestChainComponents()` to call `TestChainReader()` which will call `RunContractReaderTests()` and `RunContractReaderInterfaceTests()`

File: core/services/relay/evm/chain_components_test.go
Comment (ilija42) [inline core/services/relay/evm/chain_components_test.go]: Same as below, this needs to be more granular

File: core/services/relay/evm/chain_components_test.go
Comment (ilija42) [inline core/services/relay/evm/chain_components_test.go:293]: seems overkill to start a whole app just for keystore and chain.txm, is there simpler way to do this? I am pretty sure that you can create a keystore easily

**Review by ilija42 — COMMENTED** [review]
small nits, otherwise looks good

File: core/services/relay/evm/chain_components_test.go
Comment (silaslenihan) [reply core/services/relay/evm/chain_components_test.go:293]: Yes, keystore can be created easier as I do in the `chainlink-evm-testing` [implementation](https://github.com/smartcontractkit/chainlink-evm-testing/blob/d58f97fdceae2d3800878146ee9b14e68d362bb2/chainlink-dev/plugins/evm.interface-test/pkg/chain_components_tester.go#L117-L120), however, I'm not sure if there's an easier way to start up a TXM without the underlying simulated app / chain, but I can look into this.

File: core/services/relay/evm/chain_components_test.go
Comment (silaslenihan) [reply core/services/relay/evm/chain_components_test.go]: I reverted these tests to be specific to `ContractReader`

File: core/services/relay/evm/evmtesting/chain_writer_historical_wrapper.go
Comment (nolag) [inline core/services/relay/evm/evmtesting/chain_writer_historical_wrapper.go]: What test(s) use historical data? 

I feel like this shouldn't be reusable, if we need this ability to assure compatibility, it's ok to mock it for the simulated backend but not for a real chain.  Therefore, I would put it in evm/ and end with _test so it's only usable by that test suite.

Comment (silaslenihan) [reply core/services/relay/evm/evmtesting/chain_writer_historical_wrapper.go]: > What test(s) use historical data?

The ones which test the `ChainReader`'s confidence level based reading I believe. But you're right, these tests _are_ only use by the Simulated Backend and thus can be moved to testing files in `/evm`

Comment (silaslenihan) [reply core/services/relay/evm/evmtesting/chain_writer_historical_wrapper.go]: Moved the two wrappers two `/evm` with `_test` suffixes

**Review by ilija42 — DISMISSED** [review]

File: core/services/relay/evm/chain_components_test.go
Comment (patrick-dowell) [reply core/services/relay/evm/chain_components_test.go:293]: > Yes, keystore can be created easier as I do in the chainlink-evm-testing [implementation](https://github.com/smartcontractkit/chainlink-evm-testing/blob/d58f97fdceae2d3800878146ee9b14e68d362bb2/chainlink-dev/plugins/evm.interface-test/pkg/chain_components_tester.go#L117-L120), however, I'm not sure if there's an easier way to start up a TXM without the underlying simulated app / chain, but I can look into this.

Has this been resolved?

**Review by patrick-dowell — DISMISSED** [review]
A lot of this is around the setup of these tests and is pretty specific to how these tests work and how you need to initialize the TXM and CW properly. Overall it looks reasonable but I'd want Ilija's approval for sure before merging. I know there are parts of this that I don't fully understand.

**Review by ilija42 — APPROVED** [review]

**Review by ilija42 — CHANGES_REQUESTED** [review]
whats up with the commit history?

File: core/services/relay/evm/chain_components_test.go
Comment (ilija42) [reply core/services/relay/evm/chain_components_test.go:293]: I see txm being initialized on its own in tests in three places [1](https://github.com/smartcontractkit/chainlink/blob/9ce13a13a1c03cdab26896e37b25c9e247ef6095/core/chains/evm/txmgr/evm_tx_store_test.go#L1367), [2](https://github.com/smartcontractkit/chainlink/blob/9ce13a13a1c03cdab26896e37b25c9e247ef6095/core/services/vrf/v2/integration_v2_test.go#L144) , [3](https://github.com/smartcontractkit/chainlink/blob/9ce13a13a1c03cdab26896e37b25c9e247ef6095/core/services/vrf/v2/listener_v2_test.go#L42), can we reuse any of these approaches so that we don't have to init the whole app?

**Comment by silaslenihan** [comment]
> whats up with the commit history?

I believe Amit's commits are included in this PR because I was dependent on the Finazlier changes he was working on. While those were still in progress I based this branch on his branch. I'm not sure why they are still there though because I rebased on top of develop

**Comment by ilija42** [comment]
> > whats up with the commit history?
> 
> I believe Amit's commits are included in this PR because I was dependent on the Finazlier changes he was working on. While those were still in progress I based this branch on his branch. I'm not sure why they are still there though because I rebased on top of develop

HIs commits got squashed so when you rebase it doesn't recognize them

**Review by ilija42 — COMMENTED** [review]
You are effectively duplicating commit history although it will squash, but your squashed commit msg won't reflect the actual changes

File: core/services/relay/evm/chain_components_test.go
Comment (silaslenihan) [reply core/services/relay/evm/chain_components_test.go:293]: The issue with these examples is they don't implement the underlying components of the TXM (i.e. Broadcaster, confirmer, finalizer, etc.) that are necessary to actually send transactions (they just leave them `nil`).  There may be a way to individually set up these underlying components, but I am not sure it is worth the effort

**Review by EasterTheBunny — DISMISSED** [review]

**Review by EasterTheBunny — APPROVED** [review]

**Review by ilija42 — APPROVED** [review]

**Pull request merged**
